File: 2024-2025/1.introduction_i-o_data_structures/barbarian.cpp
// NOTE: it is recommended to use this even if you don't understand the following code.

#include <fstream>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int main()
{
    // uncomment the two following lines if you want to read/write from files
    ifstream cin("input.txt");
    // ofstream cout("output.txt");

    int N;
    cin >> N;

    vector<int> D(N);
    for (int i = 0; i < N; ++i)
        cin >> D[i];

    for (int i = 1; i < N; ++i)
        if (D[i - 1] > D[i])
            throw invalid_argument("overflow");

    vector<int> ultima(N);

    // INSERT YOUR CODE HERE
    for (int partenza = 0; partenza < N; partenza++)
    {
        vector<bool> distrutta(N, false);
        int attuale = partenza;
        int attuale_precedente;

        for (int intere = N; intere > 0; intere--)
        {
            attuale_precedente = attuale;
            distrutta[attuale] = true;
            int prima = attuale - 1;
            while (prima >= 0 && distrutta[prima])
                prima--;
            int dopo = attuale + 1;
            while (dopo < N && distrutta[dopo])
                dopo++;

            int distanza_prima = -1;
            if (prima >= 0)
                distanza_prima = D[attuale] - D[prima];
            int distanza_dopo = -1;
            if (dopo < N)
                distanza_dopo = D[dopo] - D[attuale];

            if (distanza_prima < 0)
                attuale = dopo;
            else if (distanza_dopo < 0)
                attuale = prima;
            else if (distanza_prima <= distanza_dopo)
                attuale = prima;
            else
                attuale = dopo;
        }

        ultima[partenza] = attuale_precedente;
    }

    for (int i = 0; i < N; ++i)
        cout << ultima[i] << " ";
    cout << endl;

    return 0;
}

File: 2024-2025/1.introduction_i-o_data_structures/barbarian.py
import sys

# Redirect input from file
# sys.stdin = open('input.txt', 'r')
# sys.stdout = open('output.txt', 'w')

N = int(input())
distanze = input().split()
D = []
for i in range(N):
    D.append(int(distanze[i]))

# Controlla che il vettore sia ordinato
for i in range(1, N):
    if D[i - 1] > D[i]:
        print("Errore: overflow of input")
        sys.exit(1)

ultima = []

# Per ogni villaggio di partenza
for partenza in range(N):
    distrutta = []
    for i in range(N):
        distrutta.append(False)

    attuale = partenza
    attuale_precedente = 0

    # Distruggi N villaggi
    for intere in range(N, 0, -1):
        attuale_precedente = attuale
        distrutta[attuale] = True

        # Trova il villaggio prima non distrutto
        prima = attuale - 1
        while prima >= 0 and distrutta[prima]:
            prima -= 1

        # Trova il villaggio dopo non distrutto
        dopo = attuale + 1
        while dopo < N and distrutta[dopo]:
            dopo += 1

        # Calcola distanze
        infinito = 10**18
        distanza_prima = infinito
        if prima >= 0:
            distanza_prima = D[attuale] - D[prima]

        distanza_dopo = infinito
        if dopo < N:
            distanza_dopo = D[dopo] - D[attuale]

        # Vai al piu' vicino (preferisci sinistra se uguale)
        if distanza_prima <= distanza_dopo:
            attuale = prima
        else:
            attuale = dopo

    ultima.append(attuale_precedente)

# Stampa risultati
for i in range(N):
    print(ultima[i], end=" ")
print()
